File: 22.generate-parentheses.cpp
class Solution {
public:
    vector<string> res;

    vector<string> generateParenthesis(int n) {
        if (n == 0) {
            return res;
        }

        string v = "";
        dfs(n, v, 0, 0);
        return res;
    }

    void dfs(int n, string &v, int l, int r) {
        if (l > n || r > n || r > l) {
            return;
        }

        if (2 * n == l+r) {
            res.push_back(v);
            return;
        }

        auto tmp = v;

        v = tmp;
        v = v + "(";
        dfs(n, v, l+1, r);

        v = tmp;
        v = v + ")";
        dfs(n, v, l, r+1);

        // if (l == n) {
        //     v = v + ")";
        //     dfs(n, v, l, r + 1);
        // } else if (r == n) {
        //     return;
        // } else if (l == r) {
        //     v = v + "(";
        //     dfs(n, v, l+1, r);
        // } else {
        //     auto tmp = v;

        //     v = tmp;
        //     v = v + "(";
        //     dfs(n, v, l+1, r);

        //     v = tmp;
        //     v = v + ")";
        //     dfs(n, v, l, r+1);
        // }
    }
};

File: 22.generate-parentheses.php
class Solution {

    private $res;

    /**
     * @param Integer $n
     * @return String[]
     */
    function generateParenthesis($n) {
        $this->dfs($n, 0, 0, "");
        return $this->res;
    }
    
    function dfs($n, $l, $r, $v)
    {
        if ($l > $n) {
            return;
        }
        
        if ($r == $n) {
            $this->res[] = $v;
            return;
        }
        
        if ($l == $r) {
            $this->dfs($n, $l+1, $r, $v . "(");
            return;
        }
        
        $this->dfs($n, $l+1, $r, $v . "(");
        $this->dfs($n, $l, $r+1, $v . ")");
    }
}
